marlowe-playground-client/src/Monaco.purs

```diff
@@ -0,0 +1,310 @@
+import Foreign (unsafeToForeign)
+import Foreign.Generic (class Encode, Foreign, SumEncoding(..), defaultOptions, encode, genericEncode)
+import Foreign.Object (Object)
+import Foreign.Object as Object
+import Web.HTML (HTMLElement)
+
+class Default a where
+  default :: a
+
+newtype LanguageExtensionPoint
+  = LanguageExtensionPoint { id :: String }
+
+_id :: Lens' LanguageExtensionPoint String
+_id = _Newtype <<< prop (SProxy :: SProxy "id")
+
+derive instance newtypeLanguageExtensionPoint :: Newtype LanguageExtensionPoint _
+
+derive instance genericLanguageExtensionPoint :: Generic LanguageExtensionPoint _
+
+derive newtype instance encodeLanguageExtensionPoint :: Encode LanguageExtensionPoint
+
+newtype MonarchLanguageBracket
+  = MonarchLanguageBracket { close :: String, open :: String, token :: String }
+
+derive instance newtypeMonarchLanguageBracket :: Newtype MonarchLanguageBracket _
+
+derive instance genericMonarchLanguageBracket :: Generic MonarchLanguageBracket _
+
+derive newtype instance encodeMonarchLanguageBracket :: Encode MonarchLanguageBracket
+
+data Action
+  = Action { token :: String, next :: Maybe String, log :: Maybe String }
+  | Cases { cases :: (Object String), log :: Maybe String }
+
+derive instance genericAction :: Generic Action _
+
+instance encodeAction :: Encode Action where
+  encode a =
+    let
+      sumEncoding =
+        TaggedObject
+          { tagFieldName: "tag"
+          , contentsFieldName: "contents"
+          , constructorTagTransform: identity
+          , unwrapRecords: true
+          }
+    in
+      genericEncode (defaultOptions { sumEncoding = sumEncoding }) a
+
+newtype LanguageRule
+  = LanguageRule { regex :: Regex, action :: Action }
+
+derive instance newtypeLanguageRule :: Newtype LanguageRule _
+
+derive instance genericLanguageRule :: Generic LanguageRule _
+
+instance encodeLanguageRule :: Encode LanguageRule where
+  encode (LanguageRule r) = encode { regex: unsafeToForeign r.regex, action: r.action }
+
+simpleRule :: Regex -> String -> LanguageRule
+simpleRule regex token = LanguageRule { regex, action: Action { token, next: Nothing, log: Nothing } }
+
+simpleRuleWithLog :: Regex -> String -> String -> LanguageRule
+simpleRuleWithLog regex token msg = LanguageRule { regex, action: Action { token, next: Nothing, log: Just msg } }
+
+simpleRuleWithAction :: Regex -> String -> String -> LanguageRule
+simpleRuleWithAction regex token next = LanguageRule { regex, action: Action { token, next: Just next, log: Nothing } }
+
+simpleRuleCases :: Regex -> Array (Tuple String String) -> LanguageRule
+simpleRuleCases regex cases = LanguageRule { regex, action: Cases { log: Nothing, cases: (Object.fromFoldable cases) } }
+
+simpleRuleCasesWithLog :: Regex -> String -> Array (Tuple String String) -> LanguageRule
+simpleRuleCasesWithLog regex msg cases = LanguageRule { regex, action: Cases { log: Just msg, cases: (Object.fromFoldable cases) } }
+
+newtype MonarchLanguage
```

Comment:
I'm not currently using Monarch but may use it for Haskell editor if I try to replace that. I could delete it here but it is a general Monaco purs integration.